src/poisson.f90

```diff
@@ -0,0 +1,509 @@
+!!                                                   to the position in "vb1" containing prescribed   *
+!!                                                   value of (fp).                                   *
+!!    num_side_nodes(1:4,1:num_boundary_points) .... (I1,I2,IE,IK) I1,I2 are the numbers of the nodes *
+!!                                                   on the side. IE is the element containing the    *
+!!                                                   side. IK is either 0 (no Neumann BC) or a        *
+!!                                                   pointer (IK > 0) at the position in "vb2" that   *
+!!                                                   contains the prescribed value of (q). The values *
+!!                                                   (Kx,Ky) are those corresponding to element IE.   *
+!!                                                                                                    *
+!!    FEM VARIABLES:                                                                                  *
+!!                                                                                                    *
+!!    nodal_value_of_f(1:num_nodes) ................ Nodal values of f.                               *
+!!    element_stiffness(1:6,1:num_elements) ........ Stores the entries of the element stiffness      *
+!!                                                   matrix (k11,k12,k13,k22,k23,k33).                *
+!!                                                                                                    *
+!!    PCG VARIABLES:                                                                                  *
+!!                                                                                                    *
+!!    pre_conditioning_matrix(1:num_nodes) ......... Diagonal preconditioning matrix.                 *
+!!    rhs_vector(1:num_nodes) ...................... Right-hand side (RHS) vector.                    *
+!!    f_increment(1:num_nodes) ..................... Computed increment of f.                         *
+!!    boundary_index(1:num_nodes) .................. Index to set f_increment(ip) = 0 for Dirichlet   *
+!!                                                   boundary conditions.                             *
+!!    b(1:num_nodes) ............................... Stores beta as in Hestenes-Stiefel relation.     *
+!!                                                                                                    *
+!!----------------------------------------------------------------------------------------------------*
+module poisson
+      implicit none
+
+      integer, parameter :: mxp  = 10000
+      integer, parameter :: mxe  = 30000 
+      integer, parameter :: mxb  =  5000 
+      integer, parameter :: mxc  =   100
+
+      integer :: num_nodes, num_elements, num_boundary_points, num_sets, num_dirichlet_boundary_conditions, &
+                 num_neumann_boundary_conditions
```

Comment:
I would put these into a data structure and pass into subroutines, rather than having them as globals